unity/Assets/Scripts/BaseFPSAgentController.cs

```diff
@@ -1073,16 +1075,39 @@ protected bool IsCollided()
 			return GameObject.FindObjectsOfType<SimObj>();
         }
 
+        public void ResetObjectFilter() {
+            this.simObjFilter = null;
+            actionFinished(true);
+        }
+        public void SetObjectFilter(string[] objectIds) {
+            SimObjPhysics[] simObjects = GameObject.FindObjectsOfType<SimObjPhysics>();
+            simObjFilter = new HashSet<SimObjPhysics>();
+            foreach(var simObj in simObjects) {
+                if(Array.IndexOf(objectIds, simObj.ObjectID) >= 0) {
+                    simObjFilter.Add(simObj);
+                }
+            }
+            actionFinished(true);
+        }
+
         public virtual ObjectMetadata[] generateObjectMetadata()
 		{
-            SimObjPhysics[] visibleSimObjs = VisibleSimObjs(false); // Update visibility for all sim objects for this agent
             HashSet<SimObjPhysics> visibleSimObjsHash = new HashSet<SimObjPhysics>();
-            foreach (SimObjPhysics sop in visibleSimObjs) {
-                visibleSimObjsHash.Add(sop);
+            SimObjPhysics[] simObjects = null;
+            if (this.simObjFilter != null) {
+                foreach (SimObjPhysics sop in this.simObjFilter) {
+                    if (isSimObjVisible(m_Camera, sop)) {
```

Comment:
Since you added a function that works with a single sim object, you probably don't need the `visibleSimObjsHash` in line 1125 you can replace `visibleSimObjsHash.Contains(simObj)` with a call to this new function `isSimObjVisible(m_Camera, sop)`

Reply:
If you look at the GetAllVisibleSimObjPhysics (which VisibleSimObjs calls), there is some logic there that does a spherecast first and then does a visibility check on the actual object.  This is done to avoid having to send raycasts out to every visibility point of every simobject on every frame.

Reply:
Makes sense